### ISSUE TITLE
[CURATOR-194] Deadlock in ConnectionState.checkTimeouts

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#715]
<p>When ConnectionState.checkTimeouts actually detects a timeout, it calls 'reset'  <br/>
which calls org.apache.zookeeper.ClientCnxn.close, which sends a ZooDefs.OpCode.closeSession request.<br/>
Then it waits on the packet, until SendThread calls 'notifyAll' on the packet.</p>

<p>At that time, SendThread is blocked because it tries to enter the synchronized method 'ConnectionState.checkTimeouts'.<br/>
So it will never notify the packet.</p>

<p>Here is the thread dump:</p>

<p>"job-scheduler_Worker-19-CheckHealthTask" prio=10 tid=0x00007f260609c000 nid=0x5a97 in Object.wait() <span class="error">&#91;0x00007f25723e1000&#93;</span><br/>
   java.lang.Thread.State: WAITING (on object monitor)<br/>
at java.lang.Object.wait(Native Method)</p>
<ul class="alternate" type="square">
	<li>waiting on <0x0000000725fc0580> (a org.apache.zookeeper.ClientCnxn$Packet)<br/>
at java.lang.Object.wait(Object.java:503)<br/>
at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1342)</li>
	<li>locked <0x0000000725fc0580> (a org.apache.zookeeper.ClientCnxn$Packet)<br/>
at org.apache.zookeeper.ClientCnxn.close(ClientCnxn.java:1314)<br/>
at org.apache.zookeeper.ZooKeeper.close(ZooKeeper.java:677)</li>
	<li>locked <0x0000000723949c88> (a org.apache.zookeeper.ZooKeeper)<br/>
at org.apache.curator.HandleHolder.internalClose(HandleHolder.java:139)<br/>
at org.apache.curator.HandleHolder.closeAndReset(HandleHolder.java:77)<br/>
at org.apache.curator.ConnectionState.reset(ConnectionState.java:218)</li>
	<li>locked <0x000000071651de48> (a org.apache.curator.ConnectionState)<br/>
at org.apache.curator.ConnectionState.checkTimeouts(ConnectionState.java:194)</li>
	<li>locked <0x000000071651de48> (a org.apache.curator.ConnectionState)<br/>
at org.apache.curator.ConnectionState.getZooKeeper(ConnectionState.java:88)<br/>
at org.apache.curator.CuratorZookeeperClient.getZooKeeper(CuratorZookeeperClient.java:115)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl.getZooKeeper(CuratorFrameworkImpl.java:474)<br/>
at org.apache.curator.framework.imps.ExistsBuilderImpl$2.call(ExistsBuilderImpl.java:172)<br/>
at org.apache.curator.framework.imps.ExistsBuilderImpl$2.call(ExistsBuilderImpl.java:161)<br/>
at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:107)<br/>
at org.apache.curator.framework.imps.ExistsBuilderImpl.pathInForeground(ExistsBuilderImpl.java:157)<br/>
at org.apache.curator.framework.imps.ExistsBuilderImpl.forPath(ExistsBuilderImpl.java:148)<br/>
at org.apache.curator.framework.imps.ExistsBuilderImpl.forPath(ExistsBuilderImpl.java:36)<br/>
at com.alu.dal.zooKeeper.ZooKeeperSession.checkHealth(ZooKeeperSession.java:350)<br/>
at com.alu.dal.zooKeeper.ZooKeeperSession.check(ZooKeeperSession.java:86)<br/>
at com.alu.orchestration.cluster.ClusterInstanceServiceImpl.checkQuorum(ClusterInstanceServiceImpl.java:464)<br/>
at com.alu.orchestration.cluster.ClusterInstanceServiceImpl.checkHealthState(ClusterInstanceServiceImpl.java:400)<br/>
at com.alu.tasks.health.CheckHealthTaskImpl.doWork(CheckHealthTaskImpl.java:37)<br/>
at com.alu.scheduler.JobSchedulerDetails$QuartzJob.executeInternal(JobSchedulerDetails.java:95)<br/>
at org.springframework.scheduling.quartz.QuartzJobBean.execute(QuartzJobBean.java:114)<br/>
at org.quartz.core.JobRunShell.run(JobRunShell.java:216)<br/>
at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549)</li>
</ul>

<p>"localhost-startStop-1-SendThread(11.1.1.11:2181)" daemon prio=10 tid=0x00007f257c61a000 nid=0x7c3 waiting for monitor entry <span class="error">&#91;0x00007f2562e65000&#93;</span><br/>
   java.lang.Thread.State: BLOCKED (on object monitor)<br/>
at org.apache.curator.ConnectionState.checkTimeouts(ConnectionState.java:177)</p>
<ul class="alternate" type="square">
	<li>waiting to lock <0x000000071651de48> (a org.apache.curator.ConnectionState)<br/>
at org.apache.curator.ConnectionState.getZooKeeper(ConnectionState.java:88)<br/>
at org.apache.curator.CuratorZookeeperClient.getZooKeeper(CuratorZookeeperClient.java:115)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl.performBackgroundOperation(CuratorFrameworkImpl.java:793)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl.doSyncForSuspendedConnection(CuratorFrameworkImpl.java:668)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl.access$800(CuratorFrameworkImpl.java:58)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl$7.retriesExhausted(CuratorFrameworkImpl.java:664)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl.checkBackgroundRetry(CuratorFrameworkImpl.java:683)<br/>
at org.apache.curator.framework.imps.CuratorFrameworkImpl.processBackgroundOperation(CuratorFrameworkImpl.java:496)<br/>
at org.apache.curator.framework.imps.BackgroundSyncImpl$1.processResult(BackgroundSyncImpl.java:50)<br/>
at org.apache.zookeeper.ClientCnxn$EventThread.processEvent(ClientCnxn.java:609)<br/>
at org.apache.zookeeper.ClientCnxn$EventThread.queuePacket(ClientCnxn.java:478)</li>
	<li>locked <0x0000000714935b18> (a java.util.concurrent.LinkedBlockingQueue)<br/>
at org.apache.zookeeper.ClientCnxn.finishPacket(ClientCnxn.java:630)<br/>
at org.apache.zookeeper.ClientCnxn.conLossPacket(ClientCnxn.java:648)<br/>
at org.apache.zookeeper.ClientCnxn.access$2400(ClientCnxn.java:85)<br/>
at org.apache.zookeeper.ClientCnxn$SendThread.cleanup(ClientCnxn.java:1194)</li>
	<li>locked <0x000000071b205bf0> (a java.util.LinkedList)<br/>
at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1122)</li>
</ul>

---
<details><summary><i>Originally reported by <a title="Amir Gur" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=amir.gur">amir.gur</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-194" target="_blank">Deadlock in ConnectionState.checkTimeouts</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2015-03-25T17:06:40+00:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>Possibly a duplicate of <a href="https://issues.apache.org/jira/browse/CURATOR-172" title="Deadlock when performing background operation" class="issue-link" data-issue-key="CURATOR-172"><del>CURATOR-172</del></a>? Your assessment seems reasonable though. Will try and have a look at it tomorrow in more detail.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>It's the same as <a href="https://issues.apache.org/jira/browse/CURATOR-172" title="Deadlock when performing background operation" class="issue-link" data-issue-key="CURATOR-172"><del>CURATOR-172</del></a> I'm almost certain. I'm going to close this.</p>

**Comment by jira-importer** [comment]
<i>[Duplicates: <a href="https://github.com/apache/curator/issues?q=in%3Atitle%20CURATOR-172">CURATOR-172</a>]</i>
